src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -0,0 +1,124 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.AspNetCore.Http;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class HttpContextServerCallContext : ServerCallContext
+    {
+        private readonly IWriteOptions _writeOptions;
+        private readonly Lazy<Metadata> _requestHeaders;
+        private Lazy<string> _peer;
+
+        internal HttpContextServerCallContext(HttpContext httpContext, IWriteOptions writeOptions)
+        {
+            HttpContext = httpContext;
+            _writeOptions = writeOptions;
+
+            var connection = HttpContext.Connection;
+            if (connection.RemoteIpAddress == null)
+            {
+                _peer = new Lazy<string>();
+            }
+            else
+            {
+                _peer = new Lazy<string>($"{(connection.RemoteIpAddress.AddressFamily == AddressFamily.InterNetwork ? "ipv4" : "ipv6")}:{connection.RemoteIpAddress}:{connection.RemotePort}");
+            }
+
+            _requestHeaders = new Lazy<Metadata>(() =>
+            {
+                var metadata = new Metadata();
+
+                foreach (var header in httpContext.Request.Headers)
+                {
+                    if (!header.Key.StartsWith(':'))
```

Comment:
@jtattermusch looks like there's some discrepancies between how headers are handled and I just want to call them out:
1. Metadata converts header names to lower-case whereas ASP.NET Core does not
2. ASP.NET Core exposes pseudo-headers (:path for example) in the header collection whereas Metadata doesn't allow this.
3. We do not store binary header values. We parse them as UTF-8.

Are we okay with these behavioural differences?

Reply:
> @jtattermusch looks like there's some discrepancies between how headers are handled and I just want to call them out:
> 
> 1. Metadata converts header names to lower-case whereas ASP.NET Core does not

not sure where that would be a problem. a grpc app would  access the metadata through ServerCallContext where all the entries seems to have the right casing. On the wire everything will be lowercase because it's required by HTTP/2 spec.

> 2. ASP.NET Core exposes pseudo-headers (:path for example) in the header collection whereas Metadata doesn't allow this.

Looks like that's already addressed by hiding the pseudo headers?

> 3. We do not store binary header values. We parse them as UTF-8.

When writing, the "-bin" metadata entries should just be base64 encoded.

when reading the request headers, we can recognize the "-bin" suffix and base64 decode the value when creating Metadata.Entry.

> 
> Are we okay with these behavioural differences?

Reply:
Right, I'm trying to replicate the gRPC behaviours here as much as possible but I just want to highlight them so we are aware of these differences. The only remaining point is the binary headers. We don't support these. As in, if we receive binary headers that cannot be parsed as ASCII or UTF-8. Kestrel will throw and we reject the request and sends a 400 response. I'll file a follow up issue to discuss this further as this requires a significant amount of work to support on our side.

Reply:
Filed https://github.com/grpc/grpc-dotnet/issues/71